receiver/receivercreator/factory.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/config/configerror"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+// This file implements factory for receiver_creator. A receiver_creator can create other receivers at runtime.
+
+const (
+	typeStr = "receiver_creator"
+)
+
+// Factory is the factory for receiver_creator.
+type Factory struct {
+}
+
+var _ component.ReceiverFactoryOld = (*Factory)(nil)
+
+// Type gets the type of the Receiver config created by this factory.
+func (f *Factory) Type() string {
+	return typeStr
+}
+
+// CustomUnmarshaler returns custom unmarshaler for receiver_creator config.
+func (f *Factory) CustomUnmarshaler() component.CustomUnmarshaler {
+	return func(v *viper.Viper, viperKey string, sourceViperSection *viper.Viper, intoCfg interface{}) error {
+		if sourceViperSection == nil {
+			// Nothing to do if there is no config given.
+			return nil
+		}
+		c := intoCfg.(*Config)
+
+		for subreceiverKey := range v.GetStringMap(viperKey) {
+			cfgSection := sourceViperSection.GetStringMap(subreceiverKey + "::config")
+			subreceiver, err := newSubreceiverConfig(subreceiverKey, cfgSection)
+			if err != nil {
+				return err
+			}
+
+			// Unmarshals receiver_creator configuration like rule.
+			// TODO: validate discovery rule
+			if err := sourceViperSection.UnmarshalKey(subreceiverKey, &subreceiver); err != nil {
+				return fmt.Errorf("failed to deserialize sub-receiver %q inside receiver_creator %q: %s", subreceiverKey, viperKey, err)
+			}
+
+			c.subreceiverConfigs[subreceiverKey] = subreceiver
+		}
+
+		return nil
+	}
+}
```

Comment:
Can we rely purely on the default "Unmarshaler". Maybe we can define the public "Config" as a raw object (without any custom logic like `receiverType` / `fullName` and have an internalConfig that is created from the raw Config.

Reply:
Can be done in a separate PR, maybe just file an issue to track if you want.

Reply:
Maybe I don't understand the point of CustomUnmarshaler then. We need to be able to unmarshal something like:

```yaml
receiver_creator:
   examplereceiver/1:
       rule: rule text
       config:
            receiverCustomConfig: foo
   ...
```

Is there a better way to do it? The entries under receiver_creator are arbitrary so it has to be a map but I don't think mapstructure can squash a map into the Config struct.

Reply:
I think @jrcamp is right. I don't know if there is any other way except to use the CustomUnmarshaler.